### PR TITLE
LM Format Enforcer Guided Decoding Support

### LINKED ISSUE #3713: [Feature]: Integrate with lm-format-enforcer
### 🚀 The feature, motivation and pitch

While existing Outline state machine provide great state of the art performance, it is trading off a one-off compile time when working with the schema. For endpoint products running model as a service with customers supplying many different schemas, the cost might not be acceptable. In that case, we should integrate with lm-format-enforcer from @noamgat. 

We already have an existing logits processor interface and guided decoding tested. It should be quite straightforward to add it integration for it. In the end it should be some flag choosing `--guided-decoding-backend=...`.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by noamgat**
FIX #3713
This PR implements the issue raised by simon-mo to add LMFE decoding support to the OpenAI server via a command line argument.

It introduces a new command line arg
`--guided-decoding-backend` which defaults to outlines (current implementation) but also support lm-format-enforcer.

Example curl command to test with:

If you start the server with
```
python -m vllm.entrypoints.openai.api_server --guided-decoding-backend lm-format-enforcer --model meta-llama/Llama-2-7b-chat-hf
```
You can then run
```
curl http://localhost:8000/v1/completions \
-H "Content-Type: application/json" \
-d '{
"model": "meta-llama/Llama-2-7b-chat-hf",
"prompt": "Please create a random person's name.",
"max_tokens": 100,
"temperature": 0,
"guided_json": {  "$schema": "http://json-schema.org/draft-04/schema#",  "description": "",  "type": "object",  "properties": {    "name": {      "type": "object",      "properties": {        "first": {          "type": "string",          "minLength": 1        },        "middle": {          "type": "string",          "minLength": 1        },        "last": {          "type": "string",          "minLength": 1        }      },      "required": [        "first",        "middle",        "last"      ]    }  },  "required": [    "name"  ]}
}'
```

And get a proper json response.

### CODE REVIEW

**Comment by noamgat** [comment]
The checks caused me to realize some gaps in my original PR. They have all been addressed, this PR is 100% ready to be reviewed.

File: vllm/config.py [hunk 1]
```diff
@@ -1000,6 +998,20 @@ def _get_and_verify_max_len(
     return int(max_model_len)
 
 
+@dataclass
+class DecodingConfig:
+    """Dataclass which contains the decoding strategy of the engine"""
+    guided_decoding_backend: str = 'outlines'
+    """Which guided decoding algo to use. 'outlines' / 'lm-format-enforcer'"""
```
Comment (simon-mo) [inline vllm/config.py:1005]:
```suggestion

    # Which guided decoding algo to use. 'outlines' / 'lm-format-enforcer'
    guided_decoding_backend: str = 'outlines'
```

File: vllm/config.py [hunk 2]
```diff
@@ -73,26 +73,24 @@ class ModelConfig:
...
+                 seed: int,
+                 revision: Optional[str] = None,
+                 code_revision: Optional[str] = None,
+                 tokenizer_revision: Optional[str] = None,
+                 max_model_len: Optional[int] = None,
+                 quantization: Optional[str] = None,
+                 quantization_param_path: Optional[str] = None,
+                 enforce_eager: bool = False,
+                 max_context_len_to_capture: Optional[int] = None,
+                 max_logprobs: int = 5) -> None:
```
Comment (simon-mo) [inline vllm/config.py:93]:
```suggestion
                 max_logprobs: int = 5,) -> None:
```

doesn't look like we added anything? add the trailing comma back and return formatter.

**Review by simon-mo — APPROVED** [review]
This is pretty good! However, it would be even better if you can help separate out the tests in tests/test_entrypoints/test_openai_server that tests guided decoding into a separate file and we can create a fixture over the server (perhaps different ports) to run end to end test cases. In that case we would have a lot of confidence in recommending both solutions.

**Comment by noamgat** [comment]
In order to make the tests simpler, I ended up adding a per-request option as well (same idea `guided_decoding_backend` alongside `guided_json` etc), and run the tests for both outlines and LMFE.
It required a small change to the outlines implementation, the current one modified the input tokenizer in a breaking fashion, I modified it to work on a copy of it (and added lru_cache to avoid a performance hit)

**Comment by jamestwhedbee** [comment]
Anything else you think is blocking a merge @simon-mo?

**Comment by noamgat** [comment]
The failing test is due to a temporary networking issue during the run.

```
E       socket.gaierror: [Errno -3] Temporary failure in name resolution
 ```

**Comment by noamgat** [comment]
@simon-mo any update on the review?
